File: C++/rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int sz = matrix.size();
        
        for (int i = 0; i < sz; i++) {
            for (int j = 0; j < i; j++) {
                int t = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = t;
            }
        }
        
        for (int i = 0; i < sz; i++) {
            for (int j = 0; j < sz/2; j++) {
                int t = matrix[i][j];
                matrix[i][j] = matrix[i][sz-j-1];
                matrix[i][sz-j-1] = t;
            }
        }
        
    }
};

File: Rust/rotate-image.rs
impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        let n = matrix.len();
        // transpose
        for i in 0..n {
            for j in (i+1)..n {
                (matrix[i][j], matrix[j][i]) = (matrix[j][i], matrix[i][j]);
            }
        }
        // flip horizontally
        for row in matrix {
            row.reverse();
        }
    }
}